### ISSUE TITLE
`invalid byte sequence in UTF-8` error on `cap install`

### ISSUE THREAD

**Issue by sugitak** [capistrano/capistrano#1388]
#### Problem summary

When I try running `cap install` in some conditions, I always get `invalid byte sequence in UTF-8 (ArgumentError)`.

The condition is:
- Work in a descendent directory under `~`
- `rbenv` installed under `~/.rbenv`
- Non-UTF named file created right on `~`
- Use ruby `2.1.4` or older
#### Situation

I usally work under my home directory, and I use rbenv which is installed under `~/.rbenv`.
One day, after I ran `scp` and got files to my home directory, I got into a situation that I never be able to run `cap install` again.
#### Reproducing the problem

If I create any `invalid sequence` named files in any of the ascending directories, `cap install` fails. If I delete it, it runs without any problem.

```
$ pwd
/home/sugitak/work/captest
$ echo -e "\x91" | xargs -I{} touch ~/{}
$ cap install
/home/sugitak/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/file_list.rb:401:in `glob`: invalid byte sequence in UTF-8 (ArgumentError)
        from /home/sugitak/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/file_list.rb:401:in `glob'
        from /home/sugitak/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:258:in `block in have_rakefile'
        from /home/sugitak/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:256:in `each'
        from /home/sugitak/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:256:in `have_rakefile'
        from /home/sugitak/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:659:in `find_rakefile_location'
        from /home/sugitak/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/capistrano-3.3.5/lib/capistrano/application.rb:66:in `display_error_message'
        from /home/sugitak/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:185:in `rescue in standard_exception_handling'
        from /home/sugitak/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:176:in `standard_exception_handling'
        from /home/sugitak/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:75:in `run'
        from /home/sugitak/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/capistrano-3.3.5/lib/capistrano/application.rb:15:in `run'
        from /home/sugitak/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/capistrano-3.3.5/bin/cap:3:in `<top (required)>'
        from /home/sugitak/.rbenv/versions/2.1.4/bin/cap:23:in `load'
        from /home/sugitak/.rbenv/versions/2.1.4/bin/cap:23:in `<main>'
$ echo -e "\x91" | xargs -I{} rm ~/{}
$ cap install
mkdir -p config/deploy
create config/deploy.rb
create config/deploy/staging.rb
create config/deploy/production.rb
mkdir -p lib/capistrano/tasks
create Capfile
Capified
$
```
#### Environments
- Ruby version ~2.1.4  (tried on: 2.1.4, 2.0.0-p195, 1.9.3-p429)
  - Fixed in ruby 2.2.0~
- capistrano version 3.3.5
- x86_64-linux (Scientific Linux 6.6)
- `LANG=en_US.utf8`
#### Comments

This seems to be caused by a ruby bug, whose bugfix is not backported to Ruby 2.1 or older.
The problem in capistrano, however, is in the `Capistrano::Application#capfile`. To be specific, full-path in @rakefiles cause this problem.

I'd be reporting this issue to ruby-current, though, if there is any way to take care in capistrano, I'd love to know.
#### Appendix: The original ruby problem

For ruby 2.1.4 or older, running `Dir.glob(filename, File::FNM_CASEFOLD)` against the specified directory fails, if there's any `invalid sequence` named files in any of the directories.
`Dir.glob(filename, File::FNM_CASEFOLD)` is the way `rake` looks up for its `Rakefile`s.

```
irb(main):001:0> File.write("/tmp/\x91", 1)
=> 1
irb(main):002:0> Dir.glob("/tmp/a/b", File::FNM_CASEFOLD)
ArgumentError: invalid byte sequence in UTF-8
        from (irb):2:in `glob'
        from (irb):2
        from /home/sugitak/.rbenv/versions/2.1.4/bin/irb:11:in `<main>'
irb(main):003:0> File.delete("/tmp/\x91")
=> 1
irb(main):004:0> Dir.glob("/tmp/a/b", File::FNM_CASEFOLD)
=> []
```

**Comment by mattbrictson** [comment]
Can you try this?

```
cap install --nosearch
```

That should prevent rake's default behavior of searching parent directories.

**Comment by sugitak** [comment]
I tried for the option to find out it wasn't the correct answer to the situation, but I really want to thank you @mattbrictson, I didn't know about this option, and sorry for leaving your comment so long.

In https://github.com/capistrano/capistrano/blob/v3.4.0/lib/capistrano/application.rb#L6 the `@rakefiles` will be including `capfile`, which is a method that gives us a full-pathed file name.
This full-pathed file name will be handled to rake, and rake looks for its filename as glob.
The glob search is run by ruby, and ruby searches parend directories, and goes for error.

So, in short,
```
Dir.glob("/home/sugitak/.rbenv/versions/2.1.6/lib/ruby/gems/2.1.0/gems/capistrano-3.4.0/lib/Capfile", File::FNM_CASEFOLD)
```
this causes error.

**Comment by will-in-wi** [comment]
Is this still an issue with recent Ruby?

**Comment by leehambley** [comment]
Report is 2.5yrs old and references an EOLed Ruby (April 2017,
https://www.ruby-lang.org/en/news/2017/04/01/support-of-ruby-2-1-has-ended/).
OP mentioned that newer rubies have a fix that was simply never backported.

I vote to close.
